drivers/sensor/bosch/bmp180/bmp180.c

```diff
@@ -0,0 +1,463 @@
+/* Bosch BMP180 pressure sensor
+ *
+ * Copyright (c) 2024 , Chris Ruehl
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Datasheet:
+ * https://www.mouser.hk/datasheet/2/783/BST-BMP180-DS000-1509579.pdf
+ */
+
+#include <math.h>
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/pm/device.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/sys/util.h>
+
+#include "bmp180.h"
+
+#define DT_DRV_COMPAT bosch_bmp180
+
+LOG_MODULE_REGISTER(BMP180, CONFIG_SENSOR_LOG_LEVEL);
+
+/* Calibration Registers structure */
+struct bmp180_cal_data {
+	int16_t ac1;
+	int16_t ac2;
+	int16_t ac3;
+	uint16_t ac4;
+	uint16_t ac5;
+	uint16_t ac6;
+	int16_t b1;
+	int16_t b2;
+	int16_t mb;
+	int16_t mc;
+	int16_t md;
+} __packed;
+
+struct bmp180_config {
+	const struct i2c_dt_spec i2c;
+};
+
+struct bmp180_data {
+	uint8_t osr_pressure;
+	int32_t raw_press;
+	int32_t raw_temp;
+	int32_t comp_temp;
+	struct bmp180_cal_data cal;
+};
+
+static inline int bmp180_bus_check(const struct device *dev)
+{
+	const struct bmp180_config *cfg = dev->config;
+
+	return i2c_is_ready_dt(&cfg->i2c) ? 0 : -ENODEV;
+}
+
+static inline int bmp180_reg_read(const struct device *dev, uint8_t start,
+				 uint8_t *buf, int size)
+{
+	const struct bmp180_config *cfg = dev->config;
+
+	return i2c_burst_read_dt(&cfg->i2c, start, buf, size);
+}
+
+static inline int bmp180_reg_write(const struct device *dev, uint8_t reg,
+				   uint8_t val)
+{
+	const struct bmp180_config *cfg = dev->config;
+
+	return i2c_reg_write_byte_dt(&cfg->i2c, reg, val);
+}
+
+#ifdef CONFIG_BMP180_OSR_RUNTIME
+static int bmp180_attr_set_oversampling(const struct device *dev,
+					enum sensor_channel chan, uint16_t val)
+{
+	struct bmp180_data *data = dev->data;
+
+	/* Value must be a positive value 0-3 */
+	if ((chan != SENSOR_CHAN_PRESS) || (val > 3)) {
+		return -EINVAL;
+	}
+
+	data->osr_pressure = val;
+
+	return 0;
+}
+#endif /* CONFIG_BMP180_OSR_RUNTIME */
+
+static int bmp180_attr_set(const struct device *dev, enum sensor_channel chan,
+			   enum sensor_attribute attr,
+			   const struct sensor_value *val)
+{
+	int ret;
+
+#ifdef CONFIG_PM_DEVICE
+	enum pm_device_state state;
+
+	(void)pm_device_state_get(dev, &state);
+	if (state != PM_DEVICE_STATE_ACTIVE) {
+		return -EBUSY;
+	}
+#endif /* CONFIG_PM_DEVICE */
+
+	switch (attr) {
+#ifdef CONFIG_BMP180_OSR_RUNTIME
+	case SENSOR_ATTR_OVERSAMPLING:
```

Comment:
This is a little odd to me, since `SENSOR_ATTR_OVERSAMPLING` isn't behing a kconfig, it would be odd that the driver would sometimes support this attribute. Is there a reason for not having it always enabled?

Reply:
I cannot think about an application with the need to change the oversampling rate after setup/init, more likely this is for testing. If not supported by default, the binary is bit smaller - that all.

Reply:
Gotcha, I'm OK with this then.